### ISSUE TITLE
PostgreSQL: SERIAL PRIMARY KEY is optional

### ISSUE THREAD

**Issue by kurochin143** [sequelize/sequelize-auto#543]
It seems that defaultValue for SERIAL is truthy, causing SERIAL PRIMARY KEY to be optional.

**Comment by steveschmitt** [comment]
I'm not sure what this means.  What sequelize-auto output are you getting, and what should you be getting?

**Comment by kurochin143** [comment]
generated code
```
// current output
export class Users extends Model<UsersAttributes, UsersCreationAttributes> implements UsersAttributes {
	id?: number; // optional
}

// expected output
export class Users extends Model<UsersAttributes, UsersCreationAttributes> implements UsersAttributes {
	id!: number; // required
}
```
Looking through the sequelize-auto source code. This seems to be the problem
```
  private getTypeScriptFieldOptional(table: string, field: string) {
    const fieldObj = this.tables[table][field];
    // fieldObj.defaultValue is truthy if column type is SERIAL. Causing fieldObj.allowNull to be ignored.
    return fieldObj.allowNull || (fieldObj.defaultValue || fieldObj.defaultValue === "");
  }
```
The problem also occurs in columns with DEFAULT.
It would be better if `fieldObj.defaultValue` is not considered when generating class

**Comment by steveschmitt** [comment]
I thought that **SERIAL** means that the value is generated in the database, so it is not required by the client.  Likewise, if the database provides a **defaultValue**, then the client does not have to provide it.

**Comment by kurochin143** [comment]
Yes, it's correct that SERIAL field has a default value but the generated field shouldn't be optional if the field is not nullable.
When using the Users class as a type, it doesn't make sense that I have to check whether id exist or not even though it's guaranteed to be not null.

```
const user: Users = getUser();
const other: Other = {
   userId: user.id // error here without ! because user.id is optional and userId is not
};
```

So I'm suggesting that the generated `class` to only consider not null when checking if the field is optional or not.
And `CreationAttributes` fields be optional if defaultValue is true.

Something like this

current output
```
export interface UsersAttributes {
	id?: number; // note, because SERIAL
	createdAt?: Date; // note, because DEFAULT
	username: string;
	password: string;
}

export type UsersPk = "id";
export type UsersId = Users[UsersPk];
export type UsersCreationAttributes = Optional<UsersAttributes, UsersPk>; // note

export class Users extends Model<UsersAttributes, UsersCreationAttributes> implements UsersAttributes {
	id?: number; // note
	createdAt?: Date; // note
	username!: string;
	password!: string;
```

expected output
```
export interface UsersAttributes {
	id: number; // note
	createdAt: Date; // note
	username: string;
	password: string;
}

export type UsersPk = "id";
export type UsersId = Users[UsersPk];
export type UsersCreationAttributes = Optional<UsersAttributes, UsersPk | "createdAt">; // because they have default value

export class Users extends Model<UsersAttributes, UsersCreationAttributes> implements UsersAttributes {
	id!: number; // even though it has default value
	createdAt!: Date; // even though it has default value
	username!: string;
	password!: string;
```

**Comment by steveschmitt** [comment]
So if I create a new User object, I have to provide a value for `id`, even though that value is meaningless and will be replaced by a database-generated value?

**Comment by kurochin143** [comment]
No, the `Users.create()` function uses `UsersCreationAttributes`, which has id as optional field.

**Comment by steveschmitt** [comment]
Let's keep this issue open until the fix is actually published in npm.

**Comment by steveschmitt** [comment]
Fixed in 0.8.4
